xonsh/procs/executables.py

```diff
@@ -0,0 +1,112 @@
+"""Interfaces to locate executable files on file system."""
+
+import os
+from pathlib import Path
+
+from xonsh.built_ins import XSH
+from xonsh.platform import ON_WINDOWS
+
+
+def get_possible_names(name, env=None):
+    """Expand name to all possible variants based on `PATHEXT`.
+
+    PATHEXT is a Windows convention containing extensions to be
+    considered when searching for an executable file.
+
+    Conserves order of any extensions found and gives precedence
+    to the bare name.
+    """
+    env = env if env is not None else XSH.env
+    env_pathext = env.get("PATHEXT", [])
+    if not env_pathext:
+        return [name]
+    upper = name.upper() == name
+    extensions = [""] + env_pathext
+    return [name + (ext.upper() if upper else ext.lower()) for ext in extensions]
```

Comment:
I don't care for the case manipulation here. It feels like extra complexity, though I guess I see why you might do it (to retain some sort of consistency). I'm a little sad to see how much more complicated it is than the previous implementation (which was two lines). What about this instead:

```suggestion
    env = env if env is not None else XSH.env
    extensions = [""] + env.get("PATHEXT", [])
    return [name + match_case(ext, name) for ext in extensions]
```

where

```python
def match_case(ext, name):
    return ext.upper() if name.upper() == name else ext.lower()
```

Or if you really need the efficiency of only checking `name.upper() == name` once:

```suggestion
    env = env if env is not None else XSH.env
    extensions = [""] + env.get("PATHEXT", [])
    matcher = str.upper if name.upper() == name else str.lower
    return [name + matcher(ext) for ext in extensions]
```